Fix a bug where a socket connection to "bsv.aftrek.org" is successful…

AustEcon opened the pull request:
…ly established but on protocol negotiation, the result is "None"

- This crashes the entire asyncio event loop with "TypeError: cannot unpack non-iterable NoneType object"

Context:

This is the relevant traceback:

```
2023-08-10 12:43:42,021:ERROR:async:async task raised an unhandled exception
Traceback (most recent call last):
  File ".../electrumsv/electrumsv/async_.py", line 94, in _collect
    future.result()
  File ".../.pyenv/versions/3.9.16/lib/python3.9/concurrent/futures/_base.py", line 439, in result
    return self.__get_result()
  File ".../.pyenv/versions/3.9.16/lib/python3.9/concurrent/futures/_base.py", line 391, in __get_result
    raise self._exception
  File ".../electrumsv/electrumsv/network.py", line 944, in _main_task
    await group.spawn(self._monitor_wallets, group)
  File ".../electrumsv/lib/python3.9/site-packages/aiorpcx/curio.py", line 255, in aexit
    await self.join()
  File ".../electrumsv/lib/python3.9/site-packages/aiorpcx/curio.py", line 213, in join
    raise task.exception()
  File ".../electrumsv/electrumsv/network.py", line 973, in _maintain_connections
    await group.spawn(self._maintain_connection, n)
  File ".../electrumsv/lib/python3.9/site-packages/aiorpcx/curio.py", line 255, in aexit
    await self.join()
  File ".../electrumsv/lib/python3.9/site-packages/aiorpcx/curio.py", line 213, in join
    raise task.exception()
  File ".../electrumsv/electrumsv/network.py", line 987, in _maintain_connection
    await server.connect(self, n)
  File "...electrumsv/electrumsv/network.py", line 266, in connect
    await session.run()
  File ".../electrumsv/electrumsv/network.py", line 763, in run
    await self._negotiate_protocol()
  File ".../electrumsv/electrumsv/network.py", line 463, in _negotiate_protocol
    server_string, protocol_string = await self.send_request(method, args)
TypeError: cannot unpack non-iterable NoneType object
```

This only occurs for the "bsv.aftrek.org" server and is strange because we've never encountered this behaviour from any ElectrumX server in the past.

To reproduce this in isolation sheds some light:

```
import asyncio
import aiorpcx
import ssl

async def main(host, port):
    async with aiorpcx.connect_rs(host, port, ssl=ssl.SSLContext(ssl.PROTOCOL_TLS)) as session:
        print("Connected!")
        result = await session.send_request('server.version', ("1.3.16", ["1.4", "1.4.2"]))
        print(result)

asyncio.run(main('sv3.satoshi.io', 50002))
asyncio.run(main('bsv.aftrek.org', 50002))
```
Console output:
```
Connected!
['ElectrumX 1.20.2', '1.4.2']
Connected!
None
```

So a connection is still being established to bsv.aftrek.org but for some reason protocol negotiation is failing.

AustEcon committed:
Fix a bug where a socket connection to "bsv.aftrek.org" is successfully established but on protocol negotiation, the result is "None"

- This crashes the entire asyncio event loop with "TypeError: cannot unpack non-iterable NoneType object"
- Catch the raised TypeError to avoid crashing the asyncio event loop.

rt121212121 commented on electrumsv/network.py:
```diff
@@ -460,7 +460,9 @@ async def _negotiate_protocol(self):
         args = (PACKAGE_VERSION, [ version_string(PROTOCOL_MIN), version_string(PROTOCOL_MAX) ])
         try:
-            server_string, protocol_string = await self.send_request(method, args)
+            result = await self.send_request(method, args)
```
Does this check all the possible cases? Or is it still possible for this user to poison the well with variations?

rt121212121 replied:
i.e. get rid of the unsafe `None` check and just catch `TypeError` here.

AustEcon replied:
Done.

rt121212121 commented on electrumsv/network.py:
```diff
@@ -153,7 +153,7 @@ class DisconnectSessionError(Exception):
     def __init__(self, reason, *, blacklist=False):
         super().__init__(reason)
-        self.blacklist = False
+        self.blacklist = blacklist
```
We cannot enable this unless we know it is safe. It is very likely it causes further problems in unforeseen ways. IIRC Neil disabled blacklisting.

AustEcon replied:
Ok. Have put it back to how it was to be on the safe side.

rt121212121 commented:
Blocker to releasing this is I am unable to test on Windows with the deterministic dependencies. Installing hidapi errors on cython related errored.

AustEcon committed:
Remove hardcoding of the blacklist attribute

- Servers will now be blacklisted if `DisconnectSessionError` is raised with blacklist=True

AustEcon committed:
Add unittest for various bad inputs to `protocol_tuple`

- Added type assertions to `_negotiate_protocol` handler to ensure bad responses from a misbehaving ElectrumX server are appropriately caught.
